README.md
# Lugaru HD

[![build status](https://gitlab.com/osslugaru/lugaru/badges/master/build.svg)](https://gitlab.com/osslugaru/lugaru/commits/master)

This repository holds the official development effort of the open source game
Lugaru (and in particular of its HD version) for Linux, Mac OSX and Windows.

Developments from various forks have been integrated in the same code base in
an attempt to get a unified game to further develop (e.g. with new features,
new target platforms and new campaigns) and distribute.

This is the same game as the one distributed by Wolfire Games on
[Lugaru HD's website](http://www.wolfire.com/lugaru), though this repository
contains some further developments which are not (yet) featured in the
commercial version.

## Project links

This project is community-driven and uses the following communication channels:
- GitLab repository (including issues): https://gitlab.com/osslugaru/lugaru
- Project website: https://osslugaru.gitlab.io
- Mailing list: http://icculus.org/mailman/listinfo/lugaru
- IRC channel: [#lugaru on Freenode](http://webchat.freenode.net/?channels=lugaru)

## The game

Lugaru (pronounced Loo-GAH-roo) is a cross-platform third-person action game.
The main character, Turner, is an anthropomorphic rebel bunny rabbit with
impressive combat skills. In his quest to find those responsible for
slaughtering his village, he uncovers a far-reaching conspiracy involving the
corrupt leaders of the rabbit republic and the starving wolves from a nearby
den. Turner takes it upon himself to fight against their plot and save his
fellow rabbits from slavery.

## History

Lugaru was originally developed by David Rosen of [Wolfire Games](http://www.wolfire.com)
and was [open sourced in 2010](http://blog.wolfire.com/2010/05/Lugaru-goes-open-source).
It was made cross-platform with the help of [Ryan C. Gordon](http://icculus.org).

Various forks were made at that time, and the most interesting developments
were put back together under the OSS Lugaru organization, originally on
[Google Code](https://code.google.com/p/lugaru) and then on
[Bitbucket](https://bitbucket.org/osslugaru/lugaru).

This new repository on [GitLab](https://gitlab.com/osslugaru/lugaru) is run
by the same team, and aims at revitalizing the development effort to clean
things up, ensuring the code base builds and runs fine on all supported
platforms, and easing the packaging of Lugaru in Linux distributions.

Ideally, the updated code base could also be used to update the Lugaru HD
version sold by Wolfire Games once proven better than the current commercial
builds.

## Compiling

See the `COMPILING.md` for compilation instructions on all supported
platforms.

## Licensing

The source code is distributed under the GNU General Public License version 2
or (at your option) any later version (GPLv2+).

The assets (campaigns, graphical and audio assets, etc.) in the `Data` folder
are distributed under the Creative Commons Attribution - Share Alike license,
some in version 3.0 Unported (CC-BY-SA 3.0) and others in version 4.0
International (CC-BY-SA 4.0) as described in the `CONTENT-LICENSE.txt` file.


RELEASE-NOTES.md
# OSS Lugaru Release Notes

This file contains the subsequent release notes for each version of Lugaru as
released by the OSS Lugaru team, in reverse chronological order.

The numbers preceded by a hash (e.g. #9) reference GitLab issue numbers on
the [OSS Lugaru repository](https://gitlab.com/osslugaru/lugaru).

## Version 1.2 - 11 February 2017

Second release of the open source Lugaru project! It continues the refactoring
work to make the old Lugaru code more modular and easier to maintain and
extend. A particular attention was given to the development tools (in game
editor) to improve the modability of the game.

### Highlights

- Keep track of and display the game's version number in the console and
  main menu.
- Many fixes and improvements to development tools, most of them should now
  work as described in the documentation.
- New monospace font for the console, more readable.
- Configurable campaign end message.
- Various fixes to mod levels and campaigns.
- A lot of behind-the-scenes refactoring that should help bring more
  interesting changes in future releases.

### Detailed changes

**Engine:**

- Fix displaying the localized key names instead of the QWERTY ones (#62)
- Fixed bushes appearing above ground on first level loaded
- Add a monospace font for the console text to increase legibility (#23)
- Devtools: improve handling of `map` command to avoid crashes (#23)
- Devtools: prevent the `clothes` command from saving invalid paths (#59)
  * Added `clotheslist` command to inspect player clothes
- Devtools: reviewed and fixed conflicting hotkeys and crashes (#73, #89)
- Devtools: fix regression from 1.1 in changing player type (wolf/rabbit) (#88)
  * Some bindings were changed, see `Docs/DEVTOOLS.txt` for the updated list.
- Campaign end message now configurable in the campaign definition file (#60)
- Fix long dialog lines printed out of the dialog area (#86)
- Lots of code refactoring and modernisation, fixed warnings
- Many bug fixes

**Game content:**

- Fix incorrect number of levels in the Seven Tasks campaign file (caused a
  crash on Win32) (#70)
- Fix crash in "Temple of the Wind" (sventemple) and "Snowy Grotto"
  (svengrotto) levels (#76, #79)
- Add and fix missing black fur texture for Ash in the Empire campaign (#84)

**Project:**

- Applied `clang-format` code formatting to the source, using a custom
  Mozilla-based style (#18, !14)
- Start properly versioning the game with version number displayed in the menu
  and console (#82, !15)


## Version 1.1 - 14 December 2016

First release of the open source Lugaru project! It contains most commits
made from various contributors since the open sourcing in 2010, and is
therefore more advanced than the preceding Lugaru HD release by Wolfire.

### Highlights

- Multiple campaigns support, making it easier to install and play mods!
- Those Lugaru mods are included by default as alternative to the Turner
  official campaign:
  * Temple, by Silb
  * Empire, by Jendraz
  * The Seven Tasks, by Philtron R.
- All assets (animations, textures, models, maps, sounds, etc.) of Lugaru HD
  by Wolfire are under a free and open source license (CC-BY-SA 3.0)!
- Window and input management ported from SDL 1.2 to SDL 2.0, improving the
  support for modern screen resolutions and fullscreen mode, as well as
  better input handling.

### Detailed changes

**Project:**

- Migrated the project from Bitbucket to GitLab.com, the latter being more
  interesting feature-wise and based on an open source core
- Migrated the project from mercurial to git as the latter is better known,
  and necessary as part of the GitLab migration
- Set up continuous integration (CI) and binaries deployment with gitlab-ci:
  Linux GCC, Linux Clang, Windows MinGW32, Windows MinGW64
- Removed obsolete bundled dependencies, both binaries and source code

**Engine:**

- Added support for installing and playing multiple campaigns
  * The current campaign can be changed in the level selection menu
- Migrated from SDL 1.2 to SDL 2 (#2), with in particular:
  * improved screen resolution detection and handling, especially in
    fullscreen
  * switched input method from keycodes to scancodes
- Dehardcoded game data loading, making proper system-wide installs possible
  on Linux (#21)
- Renamed debug mode to devtools and improved the corresponding documentation
  Devtools are toggled with `--devtools` switch or adding a `Devtools: 1`
  entry to the config.txt (#32)
- Added automatic progress saving upon level completion, and not only when
  closing the game.
- Use standard directories to store game config and user data (#22):
  * Linux: XDG_CONFIG_HOME (default `~/.config/lugaru`) and XDG_DATA_HOME
    (default `~/.local/share/lugaru`)
  * OSX: `~/Library/Application Support/Lugaru`
  * Windows: `%APPDATA%\Lugaru`, i.e. `C:\Users\<you>\AppData\Roaming\Lugaru`
- Add Linux installation instructions, together with desktop file, metainfo,
  icon and man page (#15, #27, #28)
- Improved command line options parsing and added `--help` output (#14)
- Fixed `nomousegrab`option (#5)
- Added Fullscreen parameter in the options GUI, and Alt+Return as hotkey to
  toggle fullscreen mode
- Increased default resolution from 640x480 to 1024x768 and bumped default
  visual details to the maximum level (#15)
- Added "stereo" anaglyph 3D display mode to play with 3D glasses
- Lots of code refactoring and modernisation, fixed warnings
- Lots of bug fixes
- Fixed compilation against modern compilers (GCC 5/6 - including MinGW - and
  Clang 3.8/3.9 so far)

**Game content:**

- All Wolfire assets are now licensed under CC-BY-SA 3.0 by David Rosen,
  making Lugaru fully free (relicensed in commit 51203b5)
- Added the Temple campaign by Silb to the main repo (licensed under CC-BY-SA
  3.0, http://forums.wolfire.com/viewtopic.php?f=7&t=1066)
- Fixed erroneous hardcoded paths to textures in some maps


Source/MD5.CC
/*
Copyright (C) 2003, 2010 - Wolfire Games

This file is part of Lugaru.

Lugaru is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  

See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
*/

// MD5.CC - source code for the C++/object oriented translation and 
//          modification of MD5.

// Translation and modification (c) 1995 by Mordechai T. Abzug 

// This translation/ modification is provided "as is," without express or 
// implied warranty of any kind.

// The translator/ modifier does not claim (1) that MD5 will do what you think 
// it does; (2) that this translation/ modification is accurate; or (3) that 
// this software is "merchantible."  (Language for this disclaimer partially 
// copied from the disclaimer below).

/* based on:

MD5C.C - RSA Data Security, Inc., MD5 message-digest algorithm
MDDRIVER.C - test driver for MD2, MD4 and MD5


Copyright (C) 1991-2, RSA Data Security, Inc. Created 1991. All
rights reserved.

License to copy and use this software is granted provided that it
is identified as the "RSA Data Security, Inc. MD5 Message-Digest
Algorithm" in all material mentioning or referencing this software
or this function.

License is also granted to make and use derivative works provided
that such works are identified as "derived from the RSA Data
Security, Inc. MD5 Message-Digest Algorithm" in all material
mentioning or referencing the derived work.

RSA Data Security, Inc. makes no representations concerning either
the merchantability of this software or the suitability of this
software for any particular purpose. It is provided "as is"
without express or implied warranty of any kind.

These notices must be retained in any copies of any part of this
documentation and/or software.

*/






#include "md5.h"

#include <assert.h>
//#include <strings.h>
#include <iostream>




// MD5 simple initialization method

MD5::MD5(){

	init();

}




// MD5 block update operation. Continues an MD5 message-digest
// operation, processing another message block, and updating the
// context.

void MD5::update (uint1 *input, uint4 input_length) {

	uint4 input_index, buffer_index;
	uint4 buffer_space;                // how much space is left in buffer

	if (finalized){  // so we can't update!
		cerr << "MD5::update:  Can't update a finalized digest!" << endl;
		return;
	}

	// Compute number of bytes mod 64
	buffer_index = (unsigned int)((count[0] >> 3) & 0x3F);

	// Update number of bits
	if (  (count[0] += ((uint4) input_length << 3))<((uint4) input_length << 3) )
		count[1]++;

	count[1] += ((uint4)input_length >> 29);


	buffer_space = 64 - buffer_index;  // how much space is left in buffer

	// Transform as many times as possible.
	if (input_length >= buffer_space) { // ie. we have enough to fill the buffer
		// fill the rest of the buffer and transform
		memcpy (buffer + buffer_index, input, buffer_space);
		transform (buffer);

		// now, transform each 64-byte piece of the input, bypassing the buffer
		for (input_index = buffer_space; input_index + 63 < input_length; 
			input_index += 64)
			transform (input+input_index);

		buffer_index = 0;  // so we can buffer remaining
	}
	else
		input_index=0;     // so we can buffer the whole input


	// and here we do the buffering:
	memcpy(buffer+buffer_index, input+input_index, input_length-input_index);
}



// MD5 update for files.
// Like above, except that it works on files (and uses above as a primitive.)

void MD5::update(FILE *file){

	unsigned char buffer[1024];
	int len;

	while (len=fread(buffer, 1, 1024, file))
		update(buffer, len);

	fclose (file);

}






// MD5 update for istreams.
// Like update for files; see above.

void MD5::update(istream& stream){

	unsigned char buffer[1024];
	int len;

	while (stream.good()){
		stream.read((char *)buffer, (long)1024); // note that return value of read is unusable.
		len=stream.gcount();
		update(buffer, len);
	}

}






// MD5 update for ifstreams.
// Like update for files; see above.

void MD5::update(ifstream& stream){

	unsigned char buffer[1024];
	int len;

	while (stream.good()){
		stream.read((char *)buffer, (long)1024); // note that return value of read is unusable.
		len=stream.gcount();
		update(buffer, len);
	}

}






// MD5 finalization. Ends an MD5 message-digest operation, writing the
// the message digest and zeroizing the context.


void MD5::finalize (){

	unsigned char bits[8];
	unsigned int index, padLen;
	static uint1 PADDING[64]={
		0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
			0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
	};

	if (finalized){
		cerr << "MD5::finalize:  Already finalized this digest!" << endl;
		return;
	}

	// Save number of bits
	encode (bits, count, 8);

	// Pad out to 56 mod 64.
	index = (uint4) ((count[0] >> 3) & 0x3f);
	padLen = (index < 56) ? (56 - index) : (120 - index);
	update (PADDING, padLen);

	// Append length (before padding)
	update (bits, 8);

	// Store state in digest
	encode (digest, state, 16);

	// Zeroize sensitive information
	memset (buffer, 0, sizeof(*buffer));

	finalized=1;

}




MD5::MD5(FILE *file){

	init();  // must be called be all constructors
	update(file);
	finalize ();
}




MD5::MD5(istream& stream){

	init();  // must called by all constructors
	update (stream);
	finalize();
}



MD5::MD5(ifstream& stream){

	init();  // must called by all constructors
	update (stream);
	finalize();
}



unsigned char *MD5::raw_digest(){

	uint1 *s = new uint1[16];

	if (!finalized){
		cerr << "MD5::raw_digest:  Can't get digest if you haven't "<<
			"finalized the digest!" <<endl;
		return ( (unsigned char*) "");
	}

	memcpy(s, digest, 16);
	return s;
}



char *MD5::hex_digest(){

	int i;
	char *s= new char[33];

	if (!finalized){
		cerr << "MD5::hex_digest:  Can't get digest if you haven't "<<
			"finalized the digest!" <<endl;
		return "";
	}

	for (i=0; i<16; i++)
		sprintf(s+i*2, "%02x", digest[i]);

	s[32]='\0';

	return s;
}





ostream& operator<<(ostream &stream, MD5 context){

	stream << context.hex_digest();
	return stream;
}




// PRIVATE METHODS:



void MD5::init(){
	finalized=0;  // we just started!

	// Nothing counted, so count=0
	count[0] = 0;
	count[1] = 0;

	// Load magic initialization constants.
	state[0] = 0x67452301;
	state[1] = 0xefcdab89;
	state[2] = 0x98badcfe;
	state[3] = 0x10325476;
}



// Constants for MD5Transform routine.
// Although we could use C++ style constants, defines are actually better,
// since they let us easily evade scope clashes.

#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21




// MD5 basic transformation. Transforms state based on block.
void MD5::transform (uint1 block[64]){

	uint4 a = state[0], b = state[1], c = state[2], d = state[3], x[16];

	decode (x, block, 64);

	assert(!finalized);  // not just a user error, since the method is private

	/* Round 1 */
	FF (a, b, c, d, x[ 0], S11, 0xd76aa478); /* 1 */
	FF (d, a, b, c, x[ 1], S12, 0xe8c7b756); /* 2 */
	FF (c, d, a, b, x[ 2], S13, 0x242070db); /* 3 */
	FF (b, c, d, a, x[ 3], S14, 0xc1bdceee); /* 4 */
	FF (a, b, c, d, x[ 4], S11, 0xf57c0faf); /* 5 */
	FF (d, a, b, c, x[ 5], S12, 0x4787c62a); /* 6 */
	FF (c, d, a, b, x[ 6], S13, 0xa8304613); /* 7 */
	FF (b, c, d, a, x[ 7], S14, 0xfd469501); /* 8 */
	FF (a, b, c, d, x[ 8], S11, 0x698098d8); /* 9 */
	FF (d, a, b, c, x[ 9], S12, 0x8b44f7af); /* 10 */
	FF (c, d, a, b, x[10], S13, 0xffff5bb1); /* 11 */
	FF (b, c, d, a, x[11], S14, 0x895cd7be); /* 12 */
	FF (a, b, c, d, x[12], S11, 0x6b901122); /* 13 */
	FF (d, a, b, c, x[13], S12, 0xfd987193); /* 14 */
	FF (c, d, a, b, x[14], S13, 0xa679438e); /* 15 */
	FF (b, c, d, a, x[15], S14, 0x49b40821); /* 16 */

	/* Round 2 */
	GG (a, b, c, d, x[ 1], S21, 0xf61e2562); /* 17 */
	GG (d, a, b, c, x[ 6], S22, 0xc040b340); /* 18 */
	GG (c, d, a, b, x[11], S23, 0x265e5a51); /* 19 */
	GG (b, c, d, a, x[ 0], S24, 0xe9b6c7aa); /* 20 */
	GG (a, b, c, d, x[ 5], S21, 0xd62f105d); /* 21 */
	GG (d, a, b, c, x[10], S22,  0x2441453); /* 22 */
	GG (c, d, a, b, x[15], S23, 0xd8a1e681); /* 23 */
	GG (b, c, d, a, x[ 4], S24, 0xe7d3fbc8); /* 24 */
	GG (a, b, c, d, x[ 9], S21, 0x21e1cde6); /* 25 */
	GG (d, a, b, c, x[14], S22, 0xc33707d6); /* 26 */
	GG (c, d, a, b, x[ 3], S23, 0xf4d50d87); /* 27 */
	GG (b, c, d, a, x[ 8], S24, 0x455a14ed); /* 28 */
	GG (a, b, c, d, x[13], S21, 0xa9e3e905); /* 29 */
	GG (d, a, b, c, x[ 2], S22, 0xfcefa3f8); /* 30 */
	GG (c, d, a, b, x[ 7], S23, 0x676f02d9); /* 31 */
	GG (b, c, d, a, x[12], S24, 0x8d2a4c8a); /* 32 */

	/* Round 3 */
	HH (a, b, c, d, x[ 5], S31, 0xfffa3942); /* 33 */
	HH (d, a, b, c, x[ 8], S32, 0x8771f681); /* 34 */
	HH (c, d, a, b, x[11], S33, 0x6d9d6122); /* 35 */
	HH (b, c, d, a, x[14], S34, 0xfde5380c); /* 36 */
	HH (a, b, c, d, x[ 1], S31, 0xa4beea44); /* 37 */
	HH (d, a, b, c, x[ 4], S32, 0x4bdecfa9); /* 38 */
	HH (c, d, a, b, x[ 7], S33, 0xf6bb4b60); /* 39 */
	HH (b, c, d, a, x[10], S34, 0xbebfbc70); /* 40 */
	HH (a, b, c, d, x[13], S31, 0x289b7ec6); /* 41 */
	HH (d, a, b, c, x[ 0], S32, 0xeaa127fa); /* 42 */
	HH (c, d, a, b, x[ 3], S33, 0xd4ef3085); /* 43 */
	HH (b, c, d, a, x[ 6], S34,  0x4881d05); /* 44 */
	HH (a, b, c, d, x[ 9], S31, 0xd9d4d039); /* 45 */
	HH (d, a, b, c, x[12], S32, 0xe6db99e5); /* 46 */
	HH (c, d, a, b, x[15], S33, 0x1fa27cf8); /* 47 */
	HH (b, c, d, a, x[ 2], S34, 0xc4ac5665); /* 48 */

	/* Round 4 */
	II (a, b, c, d, x[ 0], S41, 0xf4292244); /* 49 */
	II (d, a, b, c, x[ 7], S42, 0x432aff97); /* 50 */
	II (c, d, a, b, x[14], S43, 0xab9423a7); /* 51 */
	II (b, c, d, a, x[ 5], S44, 0xfc93a039); /* 52 */
	II (a, b, c, d, x[12], S41, 0x655b59c3); /* 53 */
	II (d, a, b, c, x[ 3], S42, 0x8f0ccc92); /* 54 */
	II (c, d, a, b, x[10], S43, 0xffeff47d); /* 55 */
	II (b, c, d, a, x[ 1], S44, 0x85845dd1); /* 56 */
	II (a, b, c, d, x[ 8], S41, 0x6fa87e4f); /* 57 */
	II (d, a, b, c, x[15], S42, 0xfe2ce6e0); /* 58 */
	II (c, d, a, b, x[ 6], S43, 0xa3014314); /* 59 */
	II (b, c, d, a, x[13], S44, 0x4e0811a1); /* 60 */
	II (a, b, c, d, x[ 4], S41, 0xf7537e82); /* 61 */
	II (d, a, b, c, x[11], S42, 0xbd3af235); /* 62 */
	II (c, d, a, b, x[ 2], S43, 0x2ad7d2bb); /* 63 */
	II (b, c, d, a, x[ 9], S44, 0xeb86d391); /* 64 */

	state[0] += a;
	state[1] += b;
	state[2] += c;
	state[3] += d;

	// Zeroize sensitive information.
	memset ( (uint1 *) x, 0, sizeof(x));

}



// Encodes input (UINT4) into output (unsigned char). Assumes len is
// a multiple of 4.
void MD5::encode (uint1 *output, uint4 *input, uint4 len) {

	unsigned int i, j;

	for (i = 0, j = 0; j < len; i++, j += 4) {
		output[j]   = (uint1)  (input[i] & 0xff);
		output[j+1] = (uint1) ((input[i] >> 8) & 0xff);
		output[j+2] = (uint1) ((input[i] >> 16) & 0xff);
		output[j+3] = (uint1) ((input[i] >> 24) & 0xff);
	}
}




// Decodes input (unsigned char) into output (UINT4). Assumes len is
// a multiple of 4.
void MD5::decode (uint4 *output, uint1 *input, uint4 len){

	unsigned int i, j;

	for (i = 0, j = 0; j < len; i++, j += 4)
		output[i] = ((uint4)input[j]) | (((uint4)input[j+1]) << 8) |
		(((uint4)input[j+2]) << 16) | (((uint4)input[j+3]) << 24);
}





// Note: Replace "for loop" with standard memcpy if possible.
void MD5::memcpy (uint1 *output, uint1 *input, uint4 len){

	unsigned int i;

	for (i = 0; i < len; i++)
		output[i] = input[i];
}



// Note: Replace "for loop" with standard memset if possible.
void MD5::memset (uint1 *output, uint1 value, uint4 len){

	unsigned int i;

	for (i = 0; i < len; i++)
		output[i] = value;
}



// ROTATE_LEFT rotates x left n bits.

inline unsigned int MD5::rotate_left  (uint4 x, uint4 n){
	return (x << n) | (x >> (32-n))  ;
}




// F, G, H and I are basic MD5 functions.

inline unsigned int MD5::F            (uint4 x, uint4 y, uint4 z){
	return (x & y) | (~x & z);
}

inline unsigned int MD5::G            (uint4 x, uint4 y, uint4 z){
	return (x & z) | (y & ~z);
}

inline unsigned int MD5::H            (uint4 x, uint4 y, uint4 z){
	return x ^ y ^ z;
}

inline unsigned int MD5::I            (uint4 x, uint4 y, uint4 z){
	return y ^ (x | ~z);
}



// FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4.
// Rotation is separate from addition to prevent recomputation.


inline void MD5::FF(uint4& a, uint4 b, uint4 c, uint4 d, uint4 x, 
					uint4  s, uint4 ac){
						a += F(b, c, d) + x + ac;
						a = rotate_left (a, s) +b;
					}

					inline void MD5::GG(uint4& a, uint4 b, uint4 c, uint4 d, uint4 x, 
						uint4 s, uint4 ac){
							a += G(b, c, d) + x + ac;
							a = rotate_left (a, s) +b;
						}

						inline void MD5::HH(uint4& a, uint4 b, uint4 c, uint4 d, uint4 x, 
							uint4 s, uint4 ac){
								a += H(b, c, d) + x + ac;
								a = rotate_left (a, s) +b;
							}

							inline void MD5::II(uint4& a, uint4 b, uint4 c, uint4 d, uint4 x, 
								uint4 s, uint4 ac){
									a += I(b, c, d) + x + ac;
									a = rotate_left (a, s) +b;
								}

